### PR TITLE
Convert the FeedAdapter to Kotlin

### PR DESCRIPTION
**Pull request by florina-muntenescu**

### CODE REVIEW

File: core/src/main/java/io/plaidapp/core/feed/FeedAdapter.kt [hunk 1]
```diff
@@ -0,0 +1,455 @@
...
+            val placeholderColors = host.resources.getIntArray(loadingColorArrayId)
+            shotLoadingPlaceholders = arrayOfNulls(placeholderColors.size)
+            for (i in placeholderColors.indices) {
+                shotLoadingPlaceholders[i] = ColorDrawable(placeholderColors[i])
+            }
+        } else {
+            shotLoadingPlaceholders = arrayOf(ColorDrawable(Color.DKGRAY))
+        }
+        val initialGifBadgeColorId = a.getResourceId(R.styleable.DribbbleFeed_initialBadgeColor, 0)
+        initialGifBadgeColor = if (initialGifBadgeColorId != 0)
```
Comment (keyboardsurfer) [inline core/src/main/java/io/plaidapp/core/feed/FeedAdapter.kt]: Missing braces for the if / else statement.

File: core/src/main/java/io/plaidapp/core/feed/FeedAdapter.kt [hunk 2]
```diff
@@ -0,0 +1,455 @@
...
+        setHasStableIds(true)
+
+        // get the dribbble shot placeholder colors & badge color from the theme
+        val a = host.obtainStyledAttributes(R.styleable.DribbbleFeed)
+        val loadingColorArrayId =
+            a.getResourceId(R.styleable.DribbbleFeed_shotLoadingPlaceholderColors, 0)
+        if (loadingColorArrayId != 0) {
+            val placeholderColors = host.resources.getIntArray(loadingColorArrayId)
+            shotLoadingPlaceholders = arrayOfNulls(placeholderColors.size)
+            for (i in placeholderColors.indices) {
```
Comment (keyboardsurfer) [inline core/src/main/java/io/plaidapp/core/feed/FeedAdapter.kt]: `placeholderColors.forEach { ... }` ?

File: core/src/main/java/io/plaidapp/core/feed/FeedAdapter.kt [hunk 3]
```diff
@@ -0,0 +1,455 @@
...
+            TYPE_PRODUCT_HUNT_POST -> createProductHuntStoryHolder(parent)
+            TYPE_LOADING_MORE -> LoadingMoreHolder(
+                layoutInflater.inflate(R.layout.infinite_loading, parent, false)
+            )
+            else -> throw IllegalStateException("Unsupported View type")
+        }
+    }
+
+    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
+        when (getItemViewType(position)) {
```
Comment (keyboardsurfer) [inline core/src/main/java/io/plaidapp/core/feed/FeedAdapter.kt]: There's a lot of assumption and nullability in here. Can we make this easier to read?

File: core/src/main/java/io/plaidapp/core/feed/FeedAdapter.kt [hunk 4]
```diff
@@ -0,0 +1,455 @@
...
+                    }
+                    return false
+                }
+
+                override fun onLoadFailed(
+                    e: GlideException?,
+                    model: Any,
+                    target: Target<Drawable>,
+                    isFirstResource: Boolean
+                ): Boolean {
```
Comment (keyboardsurfer) [inline core/src/main/java/io/plaidapp/core/feed/FeedAdapter.kt]: Could be an expression

File: core/src/main/java/io/plaidapp/core/feed/FeedAdapter.kt [hunk 5]
```diff
@@ -0,0 +1,455 @@
...
+    }
+
+    private fun getItem(position: Int): PlaidItem? {
+        return if (position < 0 || position >= items!!.size) null else items!![position]
+    }
+
+    fun getItemColumnSpan(position: Int): Int {
+        return if (getItemViewType(position) == TYPE_LOADING_MORE) {
+            columns
+        } else getItem(position)!!.colspan
```
Comment (keyboardsurfer) [inline core/src/main/java/io/plaidapp/core/feed/FeedAdapter.kt]: Braces missing on `else`

File: core/src/main/java/io/plaidapp/core/feed/FeedAdapter.kt [hunk 6]
```diff
@@ -0,0 +1,455 @@
...
+        }
+        return RecyclerView.NO_POSITION
+    }
+
+    override fun getItemCount(): Int {
+        return dataItemCount + if (showLoadingMore) 1 else 0
+    }
+
+    // temporary method until we're able to move the item setting only to Activity and ViewModels
+    fun getItems(): List<PlaidItem>? {
```
Comment (keyboardsurfer) [inline core/src/main/java/io/plaidapp/core/feed/FeedAdapter.kt]: expression?

File: core/src/main/java/io/plaidapp/core/feed/FeedAdapter.kt [hunk 7]
```diff
@@ -0,0 +1,455 @@
...
+        val loadingPos = loadingMoreItemPosition
+        showLoadingMore = false
+        notifyItemRemoved(loadingPos)
+    }
+
+    override fun getPreloadItems(position: Int): List<Shot> {
+        val item = getItem(position)
+        return if (item is Shot) {
+            listOf(item)
+        } else emptyList()
```
Comment (keyboardsurfer) [inline core/src/main/java/io/plaidapp/core/feed/FeedAdapter.kt]: Missing braces

File: core/src/main/java/io/plaidapp/core/feed/FeedAdapter.kt [hunk 8]
```diff
@@ -0,0 +1,455 @@
...
+    }
+
+    override fun getPreloadRequestBuilder(item: Shot): RequestBuilder<Drawable>? {
+        return GlideApp.with(host).load(item.images.best())
+    }
+
+    private class LoadingMoreHolder internal constructor(itemView: View) :
+        RecyclerView.ViewHolder(itemView) {
+
+        private val progress: ProgressBar = itemView as ProgressBar
```
Comment (keyboardsurfer) [inline core/src/main/java/io/plaidapp/core/feed/FeedAdapter.kt]: Explicit return value is redundant.

File: core/src/main/java/io/plaidapp/core/feed/FeedAdapter.kt [hunk 9]
```diff
@@ -0,0 +1,455 @@
...
+    fun getItemColumnSpan(position: Int): Int {
+        return if (getItemViewType(position) == TYPE_LOADING_MORE) {
+            columns
+        } else getItem(position)!!.colspan
+    }
+
+    /**
+     * Main entry point for setting items to this adapter.
+     */
+    fun setItems(newItems: List<PlaidItem>) {
```
[the remark this answers is not in the document]
Comment (florina-muntenescu) [reply core/src/main/java/io/plaidapp/core/feed/FeedAdapter.kt]: in the next PR, I will completely remove items and use `ListAdapter.setItems` (this is why the the method is already called like this)

File: app/src/main/java/io/plaidapp/ui/HomeActivity.java [hunk 10]
```diff
@@ -185,7 +185,7 @@ protected void onCreate(Bundle savedInstanceState) {
         if (savedInstanceState == null) {
             animateToolbar();
         }
-        setExitSharedElementCallback(FeedAdapter.createSharedElementReenterCallback(this));
+        setExitSharedElementCallback(FeedAdapter.Companion.createSharedElementReenterCallback(this));
```
[the remark this answers is not in the document]
Comment (florina-muntenescu) [reply app/src/main/java/io/plaidapp/ui/HomeActivity.java]: `HomeActivity` is Java.  this doesn't work

File: core/src/main/java/io/plaidapp/core/feed/FeedAdapter.kt [hunk 11]
```diff
@@ -0,0 +1,455 @@
...
+            TYPE_LOADING_MORE -> LoadingMoreHolder(
+                layoutInflater.inflate(R.layout.infinite_loading, parent, false)
+            )
+            else -> throw IllegalStateException("Unsupported View type")
+        }
+    }
+
+    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
+        when (getItemViewType(position)) {
+            TYPE_DESIGNER_NEWS_STORY -> (holder as StoryViewHolder).bind((getItem(position) as Story?)!!)
```
[the remark this answers is not in the document]
Comment (florina-muntenescu) [reply core/src/main/java/io/plaidapp/core/feed/FeedAdapter.kt]: `getItemViewType` is used in multiple places. I wouldn't modify that method.
How do you expect this method to be?

File: core/src/main/java/io/plaidapp/core/feed/FeedAdapter.kt [hunk 12]
```diff
@@ -0,0 +1,455 @@
...
+            when (item) {
+                is Story -> return TYPE_DESIGNER_NEWS_STORY
+                is Shot -> return TYPE_DRIBBBLE_SHOT
+                is Post -> return TYPE_PRODUCT_HUNT_POST
+            }
+        }
+        return TYPE_LOADING_MORE
+    }
+
+    private fun getItem(position: Int): PlaidItem? {
```
[the remark this answers is not in the document]
Comment (florina-muntenescu) [reply core/src/main/java/io/plaidapp/core/feed/FeedAdapter.kt]: When moving to `ListAdapter` we will have to fix this problem. 
I wouldn't modify the logic in this PR but rather keep it just for converting the code to Kotlin.

**Pull request merged**